### PR TITLE
Optimize DCAT harvesting on large multiple-paged catalogs

### PR DESCRIPTION
**Pull request by abulte**
This results in:
- successful preview on large multiple-paged DCAT catalogs
- fastest harvesting times on large multiple-paged DCAT catalogs

This is done through a different structure in `HarvestJob.data`. Previously we were storing the whole concatenated catalog in `HarvestJob.data.graph`, even when parsing multiple pages, and deserialising the whole catalog for each `HarvestItem`. Now we're storing a paged `HarvestJob.data.graphs` with one graph element for each page. Thus we only need to deserialise the right page when processing an `HarvestItem `.

This leads to 2x fastest overall speed for a 500 datasets catalog with five pages. I've also seen harvest process task being 3x fastest on larger catalogs.

This also introduces `HARVEST_MAX_ITEMS` setting, allowing to limit the number of harvested items per source on development platforms.

---

We could probably optimize more, for instance:
- store the relevant part of the graph for each `HarvestIem` before queuing the `harvest_job_item`, but I'm not sure how this would work DCAT-wise: could we have a small enough object? We also need to dig into `dataset_from_rdf` which probably won't be easy. We could also introduce bugs while manipulating the remote catalog, as it is we feed it straight to `dataset_from_rdf`
- create a `PageHarvestJob` with only the graph for the page stored in it, but this is heavy work since we would need to reconcile this with our current architecture and UI
- bulk process `HarvestItem` instead of one by one, but once again heavy work and impact on architecture and UI

For the preview part, we just stop iterating on pagination when we have enough objects. The drawback is that a faulty page somewhere in the catalog won't be catched when preview is ran, but I think we can live with it.

### CODE REVIEW

**Comment by maudetes** [comment]
For info, I came across [this commit](https://github.com/opendatateam/udata/commit/0dbb5e4cf0c51745fb4432034daf7b9f0313f601), that introduced the use of a single complete graph.

**Comment by abulte** [comment]
@maudetes Interesting!

> This also prevent some data loss because properly slicing a graph is difficult in RDF when you don't known by advance all node properties. This will allow to parse more triplets for each dataset and resource.

**Review by maudetes — COMMENTED** [review]
Nice! It seems to work way faster (I guess the largest the catalog, the best increase) !

What about the risk of having some information missing on the graph page?
Ex a dataset being defined on a page and its distribution on a different page.

**Comment by abulte** [comment]
> What about the risk of having some information missing on the graph page?
> Ex a dataset being defined on a page and its distribution on a different page.

Ow, do you think it's possible? Would be poor design AMHA. I guess we could fallback on the complete graph if we don't find the complete node in the page, but looking at the code it won't be easy (`dataset_from_rdf` -> `resource_from_rdf` mayhem, and then what if the resource is just missing or invalid and we parse the whole catalog for nothing ?). Maybe we can let it fail for now and see what happens in real life.

**Comment by maudetes** [comment]
>

**Comment by maudetes** [comment]
> Ow, do you think it's possible? Would be poor design IMHO.

I've said that because some DCAT catalogs have distribution defined later in the graph.
Ex with this graph https://www.pigma.org/geonetwork/srv/fre/rdf.search, Datasets are defined first, then Distribution later on.
We could imagine having a pagination that cuts between node, ignoring the reference.

But indeed, let's see if it actually happens.
Are you aware of paginated DCAT that we could check? I'm thinking only of the geocatalogue

**Comment by abulte** [comment]
> Ex with this graph https://www.pigma.org/geonetwork/srv/fre/rdf.search, Datasets are defined first, then Distribution later on.

It's the same on Geocatalogue, and it's truncated properly https://www.geocatalogue.fr/www/dcat/154e240559a401aa0c8b3f03dc5836dd/catalog_5.rdf.

> Are you aware of paginated DCAT that we could check? I'm thinking only of the geocatalogue

Looking 👀

**Comment by abulte** [comment]
@maudetes So here's another paginated one: https://data.gov.ie/catalog.xml (anecdoticaly it fails because the whole catalog — 150k datasets can't be stored in mongo). The distribution is embedded in the dataset so no problem with this kind of pagination (if the catalog is not one of a medium-sized country 😅).

File: udata/harvest/tasks.py [hunk 1]
```diff
@@ -22,10 +22,11 @@ def harvest(self, ident):
     if items > 0:
         finalize = harvest_job_finalize.s(backend.job.id)
         items = [
-            harvest_job_item.s(backend.job.id, item.remote_id)
+            (backend.job.id, item.remote_id)
             for item in backend.job.items
         ]
-        chord(items)(finalize)
+        chunks = harvest_job_item.chunks(items, 10)
```
Comment (maudetes) [inline udata/harvest/tasks.py:28]: Did you see a difference with chunks?

**Review by maudetes — APPROVED** [review]
Seems to work smoothly

File: udata/harvest/tasks.py (hunk 1 above)
Comment (abulte) [reply udata/harvest/tasks.py]: Nope (should be tested on massive scale), it was a deadend, removed.

**Pull request merged**